From f326bb0646f2362850583245a1033a78e5b91c91 Mon Sep 17 00:00:00 2001
From: Rico Ganahl <rico.ganahl@bytesatwork.ch>
Date: Tue, 1 Mar 2022 14:29:16 +0100
Subject: [PATCH] bytesatwork: imx8mm: introduce bawconfig

Configure defaults for bytesatwork tool to store production data
in EEPROM.

Signed-off-by: Rico Ganahl <rico.ganahl@bytesatwork.ch>
Signed-off-by: Simone Schwizer <simone.schwizer@bytesatwork.ch>
---
 board/bytesatwork/bytedevkit/Kconfig      |  2 +
 board/bytesatwork/bytedevkit/Makefile     |  2 +
 board/bytesatwork/bytedevkit/bytedevkit.c | 74 +++++++++++++++++++++++
 board/bytesatwork/bytedevkit/spl.c        | 55 +++++++++++++++++
 include/configs/bytedevkit.h              |  5 ++
 5 files changed, 138 insertions(+)

diff --git a/board/bytesatwork/bytedevkit/Kconfig b/board/bytesatwork/bytedevkit/Kconfig
index 5c725ca2c1..a4baae8fc1 100644
--- a/board/bytesatwork/bytedevkit/Kconfig
+++ b/board/bytesatwork/bytedevkit/Kconfig
@@ -17,4 +17,6 @@ config IMX_CONFIG
 	default "board/bytesatwork/bytedevkit/imximage-8mm-lpddr4-fspi.cfg" if IMX8MM_BYTEDEVKIT_FSPI_BOOT
 	default "arch/arm/mach-imx/imx8m/imximage.cfg"
 
+source "board/bytesatwork/baw_config/Kconfig"
+
 endif
diff --git a/board/bytesatwork/bytedevkit/Makefile b/board/bytesatwork/bytedevkit/Makefile
index 2fb468a338..de634a2b26 100644
--- a/board/bytesatwork/bytedevkit/Makefile
+++ b/board/bytesatwork/bytedevkit/Makefile
@@ -13,3 +13,5 @@ obj-y += spl.o
 obj-$(CONFIG_IMX8M_LPDDR4) += lpddr4_timing_512_1024_1536_2048.o
 obj-$(CONFIG_IMX8M_LPDDR4) += lpddr4_timing_3072_4096.o
 endif
+
+obj-y += ../baw_config/
diff --git a/board/bytesatwork/bytedevkit/bytedevkit.c b/board/bytesatwork/bytedevkit/bytedevkit.c
index 8b94795eff..61fa231b01 100644
--- a/board/bytesatwork/bytedevkit/bytedevkit.c
+++ b/board/bytesatwork/bytedevkit/bytedevkit.c
@@ -29,6 +29,7 @@
 #include <asm/mach-imx/video.h>
 #include <power/pca9450.h>
 #include <linux/delay.h>
+#include "../baw_config/baw_config_eeprom.h"
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -44,6 +45,25 @@ static const iomux_v3_cfg_t wdog_pads[] = {
 	IMX8MM_PAD_GPIO1_IO02_WDOG1_WDOG_B  | MUX_PAD_CTRL(WDOG_PAD_CTRL),
 };
 
+#define I2C_PAD_CTRL (PAD_CTL_DSE6 | PAD_CTL_HYS | PAD_CTL_PUE | PAD_CTL_PE)
+#define I2C_GPIO_MUX MUX_PAD_CTRL(I2C_PAD_CTRL)
+/* I2Cs != I2C1(index 0) need fix by manually set the SION bit. Solution from:
+ * https://community.nxp.com/t5/i-MX-Processors/iMX8MM-problems-on-early-I2C-access-from-SPL/m-p/1230261/highlight/true#M169427
+ */
+#define I2C_MUX (I2C_GPIO_MUX | ((u64)(IOMUX_CONFIG_SION) << MUX_MODE_SHIFT))
+struct i2c_pads_info i2c2_pad_info = {
+	.scl = {
+		.i2c_mode = IMX8MM_PAD_I2C2_SCL_I2C2_SCL | I2C_MUX,
+		.gpio_mode = IMX8MM_PAD_I2C2_SCL_GPIO5_IO16 | I2C_GPIO_MUX,
+		.gp = IMX_GPIO_NR(5, 16),
+	},
+	.sda = {
+		.i2c_mode = IMX8MM_PAD_I2C2_SDA_I2C2_SDA | I2C_MUX,
+		.gpio_mode = IMX8MM_PAD_I2C2_SDA_GPIO5_IO17 | I2C_GPIO_MUX,
+		.gp = IMX_GPIO_NR(5, 17),
+	},
+};
+
 #ifdef CONFIG_FSL_FSPI
 int board_qspi_init(void)
 {
@@ -65,6 +85,11 @@ int board_early_init_f(void)
 
 	init_uart_clk(1);
 
+#ifdef CONFIG_SPL_BUILD
+	/* SPL does not use DM compared to u-boot, so setup manually */
+	setup_i2c(CONFIG_BAW_CONFIG_EEPROM_BUS, SPL_I2C_SPEED,
+		  CONFIG_BAW_CONFIG_EEPROM_ADDRESS, &i2c2_pad_info);
+#endif
 	return 0;
 }
 
@@ -189,5 +214,54 @@ int board_late_init(void)
 	board_late_mmc_env_init();
 #endif
 
+#ifdef CONFIG_BAW_CONFIG_EEPROM
+	if (baw_config_eeprom_init())
+		printf("EEPROM init failed!\n");
+#endif
+
+	return 0;
+}
+
+#include <asm/armv8/mmu.h>
+extern struct mm_region *mem_map;
+
+/* based on arch/arm/mach-imx/imx8m/soc.c */
+static u8 imx8m_find_dram_entry_in_mem_map(u64 phys_addr)
+{
+	u8 i;
+
+	for (i = 0; mem_map[i].size != 0; i++)
+		if (mem_map[i].phys == phys_addr)
+			return i;
+
+	hang(); /* Entry not found, this must never happen. */
+}
+
+int board_phys_sdram_size(phys_size_t *size)
+{
+	u8 index;
+	/* NOTE: value passed from SPL (same address as set in SPL) */
+	phys_size_t *ram_size = (phys_size_t *)BAW_CONFIG_RAM_SIZE_PTR_ADDR;
+
+	index = imx8m_find_dram_entry_in_mem_map(CFG_SYS_SDRAM_BASE);
+
+	if (!size)
+		return -EINVAL;
+
+	switch (*ram_size) {
+	case SZ_512M:
+	case SZ_1G:
+	case (SZ_1G + SZ_512M):
+	case SZ_2G:
+	case (SZ_2G + SZ_1G):
+		mem_map[index].size = *ram_size;
+		*size = *ram_size;
+		break;
+	case SZ_4G:
+		return -EOPNOTSUPP;
+	default:
+		return -EINVAL;
+	}
+
 	return 0;
 }
diff --git a/board/bytesatwork/bytedevkit/spl.c b/board/bytesatwork/bytedevkit/spl.c
index e4f9a150d2..d825fa81c8 100644
--- a/board/bytesatwork/bytedevkit/spl.c
+++ b/board/bytesatwork/bytedevkit/spl.c
@@ -28,6 +28,10 @@
 #include <mmc.h>
 #include <linux/delay.h>
 
+#include "lpddr4_timings.h"
+#include "../baw_config/baw_config_eeprom.h"
+#include "../baw_config/baw_config_get.h"
+
 DECLARE_GLOBAL_DATA_PTR;
 
 int spl_board_boot_device(enum boot_device boot_dev_spl)
@@ -50,8 +54,59 @@ int spl_board_boot_device(enum boot_device boot_dev_spl)
 	}
 }
 
+/* NOTE: ddr timing config params, see arch/arm/include/asm/arch-imx8m/ddr.h */
+struct dram_timing_info dram_timing;
+
+struct dram_setup {
+	enum baw_config_ram ram;
+	const phys_size_t size;
+	const struct dram_timing_info *timing;
+};
+
 void spl_dram_init(void)
 {
+	/* NOTE: store ram_size in OCRAM to pass value from SPL to u-boot */
+	phys_size_t *ram_size = (phys_size_t *)BAW_CONFIG_RAM_SIZE_PTR_ADDR;
+	const struct dram_setup dram_setups[] = {
+		{ M6_RAM_MT53E128M32D2DS_053, SZ_512M, &dram_timing_512 },
+		{ M6_RAM_MT53E256M32D2DS_053, SZ_1G, &dram_timing_1024 },
+		{ M6_RAM_MT53E384M32D2DS_053, SZ_1G + SZ_512M,
+		  &dram_timing_1536 },
+		{ M6_RAM_MT53D512M32D2DS_053, SZ_2G, &dram_timing_2048 },
+		{ M6_RAM_MT53E768M32D4DT_053, SZ_2G + SZ_1G,
+		  &dram_timing_3072 },
+		{ M6_RAM_MT53D1024M32D4DT_053_3GB, SZ_2G + SZ_1G,
+		  &dram_timing_3072 },
+		/* fallbacks */
+		{ M6_RAM_MT53D1024M32D4DT_053, SZ_512M, &dram_timing_3072 },
+		{ 0, 0, NULL },
+	};
+	struct baw_config config;
+	bool ram_found = false;
+	int i;
+
+#ifdef CONFIG_BAW_CONFIG_EEPROM
+	if (baw_config_eeprom_init())
+		printf("EEPROM init failed!\n");
+#endif
+
+	baw_config_get(&config);
+	for (i = 0; dram_setups[i].ram > 0; i++) {
+		if (config.ram == dram_setups[i].ram) {
+			memcpy(&dram_timing, dram_setups[i].timing,
+			       sizeof(dram_timing));
+			*ram_size = dram_setups[i].size;
+			ram_found = true;
+			break;
+		}
+	}
+
+	if (!ram_found) {
+		printf("Error: Unknown RAM config, use fallback.\n");
+		memcpy(&dram_timing, &dram_timing_512, sizeof(dram_timing));
+		*ram_size = dram_setups[0].size;
+	}
+
 	ddr_init(&dram_timing);
 }
 
diff --git a/include/configs/bytedevkit.h b/include/configs/bytedevkit.h
index 25129c1956..19284361c8 100644
--- a/include/configs/bytedevkit.h
+++ b/include/configs/bytedevkit.h
@@ -13,6 +13,11 @@
 #include <asm/arch/imx-regs.h>
 #include "imx_env.h"
 
+#define BAW_CONFIG_BUILTIN_PCB		M6_PCB_REV_1_0
+#define BAW_CONFIG_BUILTIN_RAM		M6_RAM_MT53E128M32D2DS_053
+#define BAW_CONFIG_BUILTIN_FLASH	EMMC_16GB
+#define BAW_CONFIG_RAM_SIZE_PTR_ADDR	0x0091ffe0
+
 #define CFG_SYS_UBOOT_BASE	\
 	(QSPI0_AMBA_BASE + CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR * 512)
 
